test/csuite/wt6616_checkpoint_oldest_ts/main.c

```diff
@@ -0,0 +1,383 @@
+
+    conn = (WT_CONNECTION *)arg;
+
+    testutil_check(conn->open_session(conn, NULL, "isolation=snapshot", &session));
+    testutil_check(session->open_cursor(session, uri, NULL, NULL, &cursor));
+
+    /* Insert and then delete the keys until we're killed. */
+    printf("Worker thread started.\n");
+    for (oldest_ts = 0, ts = 1;; ++ts) {
+        testutil_check(__wt_snprintf(kname, sizeof(kname), KEY_FORMAT, ts));
+
+        /* Insert the same value for key and value. */
+        testutil_check(session->begin_transaction(session, NULL));
+        cursor->set_key(cursor, kname);
+        data.data = kname;
+        data.size = sizeof(kname);
+        cursor->set_value(cursor, &data);
+        testutil_check(cursor->insert(cursor));
+        testutil_check(__wt_snprintf(tscfg, sizeof(tscfg), "commit_timestamp=%" PRIx64, ts));
+        testutil_check(session->commit_transaction(session, tscfg));
+
+        /* Update stable timestamp to the current timestamp. */
+        testutil_check(__wt_snprintf(tscfg, sizeof(tscfg), "stable_timestamp=%" PRIx64, ts));
+        testutil_check(conn->set_timestamp(conn, tscfg));
+
+        /* Remove the key using a higher timestamp. */
+        testutil_check(session->begin_transaction(session, NULL));
+        cursor->set_key(cursor, kname);
+        testutil_check(cursor->remove(cursor));
+        testutil_check(__wt_snprintf(tscfg, sizeof(tscfg), "commit_timestamp=%" PRIx64, ts + 1));
+        testutil_check(session->commit_transaction(session, tscfg));
+
+        /* Set the oldest timestamp to make half of the data obsolete. */
+        if (ts - oldest_ts > MAX_DATA) {
+            oldest_ts = ts - (ts - oldest_ts) / 2;
+            testutil_check(
+              __wt_snprintf(tscfg, sizeof(tscfg), "oldest_timestamp=%" PRIx64, oldest_ts));
+            testutil_check(conn->set_timestamp(conn, tscfg));
+        }
+    }
+    /* NOTREACHED */
+}
+
+/*
+ * Child process creates the database and table, and then creates the worker thread to add data
+ * until it is killed by the parent.
+ */
+static void run_workload(void) WT_GCC_FUNC_DECL_ATTRIBUTE((noreturn));
+static void
+run_workload(void)
+{
+    WT_CONNECTION *conn;
+    WT_SESSION *session;
+    wt_thread_t *thr;
+    uint32_t i;
+    char envconf[512];
+
+    thr = dcalloc(2, sizeof(*thr));
+
+    if (chdir(home) != 0)
+        testutil_die(errno, "Child chdir: %s", home);
+    testutil_check(__wt_snprintf(envconf, sizeof(envconf), ENV_CONFIG));
+
+    printf("wiredtiger_open configuration: %s\n", envconf);
+    testutil_check(wiredtiger_open(NULL, NULL, envconf, &conn));
+    testutil_check(conn->open_session(conn, NULL, NULL, &session));
+
+    /* Create the table. */
+    if (log_table)
+        testutil_check(
+          session->create(session, uri, "key_format=S,value_format=u,log=(enabled=true)"));
+    else
+        testutil_check(
+          session->create(session, uri, "key_format=S,value_format=u,log=(enabled=false)"));
```

Comment:
While I know that WT-6637 is open for the logged table, don't we really want both of these, all the time? It would seem that we want a logged table to simulate the oplog and a non-logged table to simulate a collection. Then the worker thread would work in both tables. I believe oplog and collection operations are part of the same transaction in MongoDB. If not, this should simulate whatever it is that MongoDB does.

Reply:
I'm with @sueloverso here - it would be best if we can simulate MongoDB expectations, and match them here within the purview of this change.

We should document the expectations around checkpoints and recovery when logging tables and using timestamps, if that hasn't been done somewhere already. My belief is that recovery of logged tables with timestamps works as MongoDB expects - but we definitely need to make sure there aren't corner cases we don't understand.

Reply:
@agorrod  @sueloverso The problem doesn't impact mongodb as mongodb 4.4 only requires the logged table to have a consistent snapshot data at last successful commit after recovery. It doesn't read previous versions of data with timestamps. I think the issue here is that we haven't decided what is the expected behavior of logged table using timestamps after recovery. For non-logged table, we want to maintain a window of data from the oldest timestamp to the stable timestamp in the checkpoint. For logged table, do we need to maintain a window of data from the oldest timestamp to the last successful commit timestamp? If we don't, there is no point testing logged table in this test as it is to ensure we can read data of any version in the maintained time window after recovery. My view is that we need to maintain the window for logged table as well, otherwise the semantic for logged table will be very weird. Though we have a stable and oldest timestamp in the checkpoint, we can't actually read the data with timestamps for logged table. If we want to fix this, we need to record timestamps in the log record and replay them with timestamps. It is a data format change and will take quite some time to get right.

Meanwhile, this ticket is blocking us from merging the rts-test-format feature branch back to develop. In that feature branch, we have some changes that we desire to go to 4.4. Therefore, I prefer it to be merged soon. For now, we can either disable the log test or remove it for now until we know what is the expected hehavior for logged tables that use timestamps.

Reply:
Could we change the verification piece of this program, with a comment describing the mongodb usage and test the logged and non-logged data differently? That is my preference. Or if a logged table has different expectations, why have the `-l` arg in this test at all? Perhaps add a comment describing this and state the case for non-logged only.